plugins/modules/ndo_l3out_bgp_peer.py

```diff
@@ -0,0 +1,934 @@
+
+                    mso_values[("privateASControls", "removeAll")] = private_as_controls.get("remove_all")
+                    mso_values[("privateASControls", "replaceWithLocalAS")] = private_as_controls.get("replace_with_local_as")
+
+                    if private_as_controls.get("remove_all") or private_as_controls.get("replace_with_local_as"):
+                        mso_values[("privateASControls", "removeExclusive")] = True
+
+            mso.sanitize(proposed_payload, collate=False, required=required_keys)
+            append_update_ops_data(ops, proposed_payload, bgp_peer_path, mso_values, mso_values_remove)
+    elif state == "absent":
+        if mso.existing:
+            ops.append(dict(op="remove", path=bgp_peer_path))
+
+    if not module.check_mode and ops:
+        mso_template.template = mso.request(mso_template.template_path, method="PATCH", data=ops)
+        match = get_bgp_peer_by_address(
+            mso_template,
+            mso_template.get_l3out_node_group(
+                node_group,
+                mso_template.get_l3out_object(uuid=l3out_uuid, name=l3out, fail_module=True).details,
+                fail_module=True,
+            ).details.get("bgpPeers", []),
+            ipv4_addr=ipv4_addr,
+            ipv6_addr=ipv6_addr,
+        )
+        if match:
+            if match.details.get("password", {}).get("ref"):
+                match.details["password"].pop("ref", None)
+
+            mso_template.update_config_with_template_and_references(match.details, reference_dict)
+            mso.existing = match.details  # When the state is present
+        else:
+            mso.existing = {}  # When the state is absent
+    elif module.check_mode and state != "query":  # When the state is present/absent with check mode
+        mso_template.update_config_with_template_and_references(mso.proposed, reference_dict)
+        mso.existing = mso.proposed if state == "present" else {}
+
+    mso.exit_json()
+
+
+def get_bgp_peer_by_address(mso_template, bgp_peers, ipv4_addr=None, ipv6_addr=None, fail_module=False):
+    if bgp_peers and (ipv4_addr or ipv6_addr):  # Query a specific object
+        kv_list = []
+        if ipv4_addr:
+            kv_list.append(KVPair("peerAddressV4", ipv4_addr))
+        if ipv6_addr:
+            kv_list.append(KVPair("peerAddressV6", ipv6_addr))
+
+        return mso_template.get_object_by_key_value_pairs("L3Out BGP Peer", bgp_peers, kv_list, fail_module)
+    return bgp_peers  # Query all objects
+
+
+def initiate_l3out_bgp_peer_mso_templates(mso, template_type, template_name, template_id):
+    key = (template_name, template_id)
+
+    # Check if the combination is unique and create an instance if needed
+    if key not in l3out_bgp_peer_instance_cache:
+        l3out_bgp_peer_instance_cache[key] = MSOTemplate(mso, template_type, template_name, template_id)
+
+    # Return the instance for this key
+    return l3out_bgp_peer_instance_cache[key]
```

Comment:
Should we define a generic functionality for template caching? I think I did caching before in some older modules for schema caching and @samiib also introduced some form of caching a while ago.

Reply:
Done!